infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStoragePodInterceptor.java

```diff
@@ -0,0 +1,154 @@
+
+    PodResource<Pod, DoneablePod> asyncStoragePodResource =
+        getAsyncStoragePodResource(namespace, workspaceId);
+
+    if (isNull(asyncStoragePodResource.get())) { // pod doesn't exist
+      return;
+    }
+
+    try {
+      deleteAsyncStoragePod(asyncStoragePodResource)
+          .get(DELETE_POD_TIMEOUT_IN_MIN, TimeUnit.MINUTES);
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new InfrastructureException(
+          "Interrupted while waiting for pod removal. " + ex.getMessage());
+    } catch (ExecutionException ex) {
+      throw new InfrastructureException(
+          "Error occurred while waiting for pod removal. " + ex.getMessage());
+    } catch (TimeoutException ex) {
+      throw new InfrastructureException("Pod removal timeout reached " + ex.getMessage());
+    }
+  }
+
+  private PodResource<Pod, DoneablePod> getAsyncStoragePodResource(
+      String workspaceId, String namespace) throws InfrastructureException {
+    return clientFactory.create(workspaceId).pods().inNamespace(namespace).withName(ASYNC_STORAGE);
+  }
+
+  private CompletableFuture<Void> deleteAsyncStoragePod(PodResource<Pod, DoneablePod> podResource)
+      throws InfrastructureException {
+    Watch toCloseOnException = null;
+    try {
+      final CompletableFuture<Void> deleteFuture = new CompletableFuture<>();
+      final Watch watch = podResource.watch(new DeleteWatcher<>(deleteFuture));
+      toCloseOnException = watch;
+
+      Boolean deleteSucceeded = podResource.withPropagationPolicy("Background").delete();
+      if (deleteSucceeded == null || !deleteSucceeded) {
+        deleteFuture.complete(null);
+      }
+      return deleteFuture.whenComplete(
+          (v, e) -> {
+            if (e != null) {
+              LOG.warn("Failed to remove pod {} cause {}", ASYNC_STORAGE, e.getMessage());
+            }
+            watch.close();
+          });
+    } catch (KubernetesClientException e) {
+      if (toCloseOnException != null) {
+        toCloseOnException.close();
+      }
+      throw new KubernetesInfrastructureException(e);
+    } catch (Exception e) {
+      if (toCloseOnException != null) {
+        toCloseOnException.close();
+      }
+      throw e;
+    }
+  }
+
+  private static class DeleteWatcher<T> implements Watcher<T> {
+
+    private final CompletableFuture<Void> future;
+
+    private DeleteWatcher(CompletableFuture<Void> future) {
+      this.future = future;
+    }
+
+    @Override
+    public void eventReceived(Action action, T hasMetadata) {
+      if (action == Action.DELETED) {
+        future.complete(null);
+      }
+    }
+
+    @Override
+    public void onClose(KubernetesClientException e) {
+      // if event about removing is received then this completion has no effect
+      future.completeExceptionally(
+          new RuntimeException(
+              "Websocket connection is closed. But event about removing is not received.", e));
```

Comment:
This is a very concrete class used in a very concrete situation, so I think the error message here should be more concrete, too.

Reply:
fixed